flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java

```diff
@@ -406,11 +403,30 @@ private void addEntryToBuffer(RequestEntryT entry, boolean insertAtHead) {
      * a failure/restart of the application.
      */
     @Override
-    public List<Collection<RequestEntryT>> snapshotState() {
-        return Arrays.asList(
-                bufferedRequestEntries.stream()
-                        .map(RequestEntryWrapper::getRequestEntry)
-                        .collect(Collectors.toList()));
+    public List<BufferedRequestState<RequestEntryT>> snapshotState() {
+        return Collections.singletonList(
+                new BufferedRequestState<>(
+                        Collections.unmodifiableCollection(bufferedRequestEntries)));
+    }
+
+    protected void initialize(BufferedRequestState<RequestEntryT> state) {
+        this.bufferedRequestEntries.clear();
+        this.bufferedRequestEntries.addAll(state.getBufferedRequestEntries());
+
+        long sum = 0L;
+        for (RequestEntryWrapper<RequestEntryT> wrapper : bufferedRequestEntries) {
+            if (wrapper.getSize() > maxRecordSizeInBytes) {
+                throw new IllegalStateException(
+                        String.format(
+                                "State contains record of size %d which exceeds sink maximum record size %d.",
+                                wrapper.getSize(), maxRecordSizeInBytes));
+            }
+
+            sum = sum + wrapper.getSize();
+        }
+
+        this.bufferedRequestEntriesTotalSizeInBytes = sum;
+        flushIfAble();
```

Comment:
Is this the right time to flush? Would this state restore be called during operator `initializing` state? If so, we probably do not want to flush here